examples/src/java/com/twitter/heron/examples/streamlet/SmartWatchTopology.java

```diff
@@ -0,0 +1,138 @@
+import com.twitter.heron.streamlet.Config;
+import com.twitter.heron.streamlet.KeyValue;
+import com.twitter.heron.streamlet.Runner;
+import com.twitter.heron.streamlet.WindowConfig;
+
+/**
+ * This topology shows an example usage of a reduce function. A source streamlet emits smart watch readings every 10
+ * seconds from one of several joggers. Those readings provide  * The processing graph then converts those smart watch
+ * readings to a KeyValue object, which is passed to a reduce function that calculates a per-jogger total number of feet
+ * run in the last minute. The reduced value is then used to provide a per-runner average pace (feet per minute) and the
+ * result is logged using a consume operation (which allows for a formatted log).
+ */
+public final class SmartWatchTopology {
+  private SmartWatchTopology() {
+  }
+
+  private static final Logger LOG =
+      Logger.getLogger(SmartWatchTopology.class.getName());
+
+  private static final List<String> JOGGERS = Arrays.asList(
+      "bill",
+      "ted"
+  );
+
+  private static class SmartWatchReading implements Serializable {
+    private static final long serialVersionUID = -6555650939020508026L;
+    private final String joggerId;
+    private final int feetRun;
+
+    SmartWatchReading() {
+      Utils.sleep(1000);
+      this.joggerId = StreamletUtils.randomFromList(JOGGERS);
+      this.feetRun = ThreadLocalRandom.current().nextInt(200, 400);
+      LOG.info(String.format("Emitted smart watch reading: %s", this));
+    }
+
+    String getJoggerId() {
+      return joggerId;
+    }
+
+    int getFeetRun() {
+      return feetRun;
+    }
+
+    @Override
+    public String toString() {
+      return String.format("(jogger: %s, feetRun: %d)", joggerId, feetRun);
+    }
+  }
+
+  public static void main(String[] args) throws Exception {
+    Builder processingGraphBuilder = Builder.createBuilder();
+
+    processingGraphBuilder.newSource(SmartWatchReading::new)
+        .setName("incoming-watch-readings")
+        .reduceByKeyAndWindow(
+            // Key extractor
+            reading -> reading.getJoggerId(),
+            // Value extractor
+            reading -> reading.getFeetRun(),
+            // The time window (1 minute of clock time)
+            WindowConfig.TumblingTimeWindow(Duration.ofSeconds(10)),
+            // The reduce function (produces a cumulative sum)
+            (cumulative, incoming) -> cumulative + incoming
+        )
+        .setName("reduce-to-total-distance-per-jogger")
+        .map(keyWindow -> {
+          // The per-key result of the previous reduce step
+          long totalFeetRun = keyWindow.getValue();
+
+          // The amount of time elapsed
+          long startTime = keyWindow.getKey().getWindow().getStartTime();
+          long endTime = keyWindow.getKey().getWindow().getEndTime();
+          long timeLengthMillis = endTime - startTime; // Cast to float to use as denominator
+
+          LOG.info(
```

Comment:
might be better to pipe this to the log operator

Reply:
It’s eventually logged in the consume operator. In this case I’m using it mostly for debugging so it can be removed.